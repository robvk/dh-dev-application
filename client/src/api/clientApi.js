import axios from "axios";

const clientSearchQuery = (searchString, limit, offset) => {
  let searchArgString = `searchString: "${searchString}"`;
  if (limit) searchArgString += `, limit: ${limit}`;
  if (offset) searchArgString += `, offset: ${offset}`;

  return `
{
  clients(${searchArgString}) {
    totalCount
    items {
        id
        first_name
        last_name
        photo
        origin
    }
  }
}
`;
};

export const findClientsAsync = (
  searchString: string,
  limit: number,
  offset: number,
  callback: Function
) =>
  axios
    .get(`/graphql?query=${clientSearchQuery(searchString, limit, offset)}`)
    .then(res => {
      const data = res.data.data.clients;
      callback(data.items, data.totalCount);
    });
