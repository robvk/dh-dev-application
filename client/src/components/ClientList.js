import React from "react";
import _ from "lodash";

import "./ClientList.css";
import ClientListItem from "./ClientListItem";
import type { Client } from "../lib/flowTypes";

type Props = {
  clients: Array<Client>,
  selectedClient: Client
};

const ClientList = (props: Props) => {
  const { clients, selectedClient } = props;
  return (
    <div className="client-dropdown-list">
      {_.map(clients, client => (
        <ClientListItem
          key={client.id}
          client={client}
          isSelected={selectedClient && client.id === selectedClient.id}
        />
      ))}
    </div>
  );
};

export default ClientList;
