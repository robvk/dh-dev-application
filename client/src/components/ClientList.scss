@import "../appColors";

.client-dropdown-list {
  z-index: 1;
  width: 100%;
  border: 1px solid $primary-color;
  border-top: 0;
  box-sizing: border-box;
}
