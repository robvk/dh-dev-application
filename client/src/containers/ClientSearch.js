import React, { Component } from "react";
import _ from "lodash";

import { findClientsAsync } from "../services/clients";

class ClientSearch extends Component {
  constructor(props) {
    super(props);

    this.state = {
      searchString: "",
      clients: []
    };

    this.inputChanged.bind(this);
    this.onReceiveData.bind(this);
  }

  onReceiveData(clientList) {
    this.setState({
      ...this.state,
      clients: clientList
    });
  }

  inputChanged(newInputString) {
    this.setState({
      ...this.state,
      searchString: newInputString
    });

    findClientsAsync(newInputString, this.onReceiveData.bind(this));
  }

  render() {
    const { searchString, clients } = this.state;

    return (
      <div>
        <input
          type="text"
          value={searchString}
          onChange={event => this.inputChanged(event.target.value)}
        />
        {_.map(clients, client => {
          return (
            <p key={client.id}>
              {client.first_name} {client.last_name} from {client.origin}
            </p>
          );
        })}
      </div>
    );
  }
}

export default ClientSearch;
