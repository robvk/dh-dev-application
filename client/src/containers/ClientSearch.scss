@import "../appColors";

.client-search {
  padding: 10px;
  position: relative;
  display: inline-block;
  width: 100%;
  box-sizing: border-box;
}

.client-search__field {
  padding: 10px;
  width: 100%;
  border: 1px solid;
  border-color: $foreground-color;
  border-radius: 3px;
  box-sizing: border-box;
}

.client-search__field:focus {
  border-color: $primary-color;
  outline: none;
}

.client-search__field--dropdown-showing {
  border-color: $primary-color;
  border-bottom-left-radius: 0px;
  border-bottom-right-radius: 0px;
  border-bottom: 0;
}
