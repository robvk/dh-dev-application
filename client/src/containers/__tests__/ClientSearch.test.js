import React from "react";
import ReactDOM from "react-dom";

jest.mock("../../api/clientApi.js");
import { findClientsAsync } from "../../api/clientApi";

import ClientSearch from "../ClientSearch";
import ClientList from "../../components/ClientList";

const manyClients = Array(15).fill({
  id: "59a2afbbfc13ae278700044e",
  first_name: "Bartram",
  last_name: "Nassi",
  email: "bnassi2@goodreads.com",
  gender: "Male",
  photo: "https://robohash.org/aeablanditiis.bmp?size=50x50&set=set1",
  origin: "Egypt"
});

findClientsAsync.mockImplementation((newInputString, callback) =>
  callback(manyClients)
);

describe("the component", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<ClientSearch />);
    expect(wrapper).toMatchSnapshot();
  });

  it("hides the dropdown when there is no input", () => {
    const wrapper = shallow(<ClientSearch />);
    expect(wrapper.find(ClientList).length).toEqual(0);
  });

  it("shows the dropdown when there is input", () => {
    const wrapper = shallow(<ClientSearch />);
    wrapper.setState({ searchString: "ab", fullClientList: [] });
    expect(wrapper.find(ClientList).length).toEqual(1);
  });

  it("only passes max 10 items to the ClientList", () => {
    const wrapper = shallow(<ClientSearch />);
    wrapper.setState({ searchString: "ab", fullClientList: manyClients });
    expect(wrapper.find(ClientList).props().clients.length).toEqual(10);
  });

  it("should set client state on receiving data", () => {
    const wrapper = shallow(<ClientSearch />);
    expect(wrapper.state("fullClientList").length).toEqual(0);
    wrapper.instance().onReceiveData(manyClients);
    expect(wrapper.state("fullClientList").length).toEqual(manyClients.length);
  });

  it("should set searchString state from input", () => {
    const wrapper = shallow(<ClientSearch />);
    expect(wrapper.state("searchString")).toEqual("");
    wrapper.find("input").simulate("change", { target: { value: "foo" } });
    expect(wrapper.state("searchString")).toEqual("foo");
  });
});

const clientSearchWithClients = () => {
  const wrapper = shallow(<ClientSearch />);
  wrapper.setState({ fullClientList: manyClients });
  return wrapper;
};

describe("when the user", () => {
  it("presses the down arrow when nothing is selected, the first item is selected", () => {
    const wrapper = clientSearchWithClients();
    expect(
      wrapper.state("fullClientList")[wrapper.state("selectedIndex")]
    ).toBeFalsy();
    wrapper.find("input").simulate("keyDown", { key: "ArrowDown" });
    expect(wrapper.state("selectedIndex")).toBe(0);
  });

  it("presses the up arrow when the first item is selected, the first item remains selected", () => {
    const wrapper = clientSearchWithClients();
    wrapper.setState({ selectedIndex: 0 });
    wrapper.find("input").simulate("keyDown", { key: "ArrowUp" });
    expect(wrapper.state("selectedIndex")).toBe(0);
  });

  it("presses the down arrow when the first item is selected, the second item is selected", () => {
    const wrapper = clientSearchWithClients();
    wrapper.setState({ selectedIndex: 0 });
    wrapper.find("input").simulate("keyDown", { key: "ArrowDown" });
    expect(wrapper.state("selectedIndex")).toBe(1);
  });

  it("presses the down arrow when the last item is selected, the last item remains selected", () => {
    const wrapper = clientSearchWithClients();
    wrapper.setState({ selectedIndex: manyClients.length });
    wrapper.find("input").simulate("keyDown", { key: "ArrowDown" });
    expect(wrapper.state("selectedIndex")).toBe(manyClients.length);
  });

  it("presses the up arrow when the second item is selected, the first item is selected", () => {
    const wrapper = clientSearchWithClients();
    wrapper.setState({ selectedIndex: 1 });
    wrapper.find("input").simulate("keyDown", { key: "ArrowUp" });
    expect(wrapper.state("selectedIndex")).toBe(0);
  });

  it("presses the down arrow when the ninth item is selected, the next page of clients are shown", () => {
    const wrapper = clientSearchWithClients();
    wrapper.setState({ searchString: "as", selectedIndex: 9 });
    expect(wrapper.instance().clientsToDisplay()[0].id).toBe(manyClients[0].id);
    wrapper.find("input").simulate("keyDown", { key: "ArrowDown" });
    expect(wrapper.instance().clientsToDisplay()[0].id).toBe(
      manyClients[10].id
    );
  });
});
