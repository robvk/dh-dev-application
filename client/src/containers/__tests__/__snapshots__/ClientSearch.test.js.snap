// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component renders correctly 1`] = `
<div
  className="client-search"
>
  <input
    className="client-search__field"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</div>
`;
